src/pools.jl

```diff
@@ -0,0 +1,109 @@
+The provided function `f` must create a new object instance of type `T`.
+The acquired object MUST be returned to the pool by calling `release(pool, key, obj)` exactly once.
+The `forcenew` keyword argument can be used to force the creation of a new object, ignoring any existing objects in the pool.
+The `isvalid` keyword argument can be used to specify a function that will be called to determine if an object is still valid
+for reuse. By default, all objects are considered valid.
+If there are no objects available for reuse, `f` will be called to create a new object.
+If the pool is already at its maximum capacity, `acquire` will block until an object is returned to the pool via `release`.
+"""
+function Base.acquire(f, pool::Pool{K, T}, key=nothing; forcenew::Bool=false, isvalid::Function=TRUE) where {K, T}
+    key isa K || throw(ArgumentError("invalid key `$key` provided for pool key type $K"))
+    acquire(pool.sem)
+    try
+        # once we've received our permit, we can figure out where the object should come from
+        # if we're forcing a new object, we can just call f() and return it
+        forcenew && return f()
+        # otherwise, check if there's an existing object in the pool to reuse
+        objs = Base.@lock pool.values get!(() -> ConcurrentStack{T}(), pool.values[], key)
+        obj = pop!(objs)
+        while obj !== nothing
+            # if the object is valid, return it
+            isvalid(obj) && return obj
+            # otherwise, try the next object
+            obj = pop!(objs)
+        end
+        # if we get here, we didn't find any valid objects, so we'll just create a new one
+        return f()
+    catch
+        # an error occurred while acquiring, so make sure we return the semaphore permit
+        release(pool.sem)
+        rethrow()
+    end
+end
+
+"""
+    release(pool::Pool{K, T}, key::K, obj::Union{T, Nothing}=nothing)
+    release(pool::Pool{K, T}, obj::T)
+    release(pool::Pool{K, T})
+
+Return an object to a `pool`, optionally keyed by the provided `key`.
+If `obj` is provided, it will be returned to the pool for reuse. Otherwise, if `nothing` is returned,
+just the "permit" will be returned to the pool.
+"""
+function Base.release(pool::Pool{K, T}, key, obj::Union{T, Nothing}=nothing) where {K, T}
+    key isa K || throw(ArgumentError("invalid key `$key` provided for pool key type $K"))
+    # if we're given an object, we'll put it back in the pool
+    # otherwise, we'll just return the permit
+    if obj !== nothing
+        # first though, we repeat Base.Semaphore's error check in the case of an invalid release
+        # where we don't want to push an object for reuse for an invalid release
+        Base.@lock pool.sem.cond_wait begin
+            if pool.sem.curr_cnt > 0
+                # if the key is invalid, we'll just let the KeyError propagate
+                objs = Base.@lock pool.values pool.values[][key]
+                push!(objs, obj)
```

Comment:
So the `push!` method allocates a new `Node` object every time (see the benchmarks shared above) which might be necessary for things to be threadsafe (?), but when `max` is provided by the user, IIUC, there is an upper bound of `Node`s that can be populated at any given time, so in that case we could reuse the `Node`s we allocated before? Like when we `pop!` a `Node`, we'd just keep it around and reuse it later?

Reply:
Hmmmmm......maybe? We'd have to make the `Node`s mutable so you could reset the values. Maybe we could pass the `max` along to the `ConcurrentStack` struct and it could create its own little pool of `Node`s that would get reused? Probably worth exploring to see if we can avoid the allocations.

Reply:
Yeah sounds good! I think getting rid of allocs could have a nice impact on performance as well. Nodes are already mutable so its really about managing the pool... which is becoming a theme here:)

Reply:
It would be nice to get rid of the allocs even when `max` is not provided so maybe a small pool could be maintained even in that case?

Reply:
There are now no allocations in `acquire`/`release` on my refactored PR